### PR TITLE
[AIRFLOW-1805] Allow to pass in Slack token through connection

### PR DESCRIPTION
**Pull request by KevinYang21**
Dear Airflow maintainers,

Please accept this PR. I understand that it will not be reviewed until I have checked off all the steps below!

### JIRA
- [x] My PR addresses the following [Airflow JIRA](https://issues.apache.org/jira/browse/AIRFLOW-1805) issues and references them in the PR title.
https://issues.apache.org/jira/browse/AIRFLOW-1805

### Description
- [x] Here are some details about my PR, including screenshots of any UI changes:
Allow users to pass in Slack token through connection which can provide better security. This enables user to expose token only to workers instead to both workers and schedulers.

### Tests
- [x] My PR adds the following unit tests:
test/hooks/test_slack_hook.py
test/operators/slack_operator.py
- [x] Tested manually with different use cases

### Commits
- [x] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"

### CODE REVIEW

**Comment by KevinYang21** [comment]
@aoen PTAL
Thank you

**Comment by Fokko** [comment]
@yrqls21 Thanks for the contribution, looks really nice. Could you rewrite your commit message according to the rules? Thanks!

File: airflow/hooks/slack_hook.py [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+        """
+        self.token = self.__get_token(token, slack_conn_id)
+
+    def __get_token(self, token, slack_conn_id):
+        if token is not None:
+            return token
+        elif slack_conn_id is not None:
+            conn = self.get_connection(slack_conn_id)
+
+            if not hasattr(conn, 'password') or not getattr(conn, 'password'):
```
Comment (aoen) [inline airflow/hooks/slack_hook.py]: Nit: can just do if not getattr(conn, 'password', None):

File: airflow/operators/slack_operator.py [hunk 2]
```diff
@@ -36,12 +38,23 @@ class SlackAPIOperator(BaseOperator):
...
+                 method=None,
                  api_params=None,
                  *args, **kwargs):
         super(SlackAPIOperator, self).__init__(*args, **kwargs)
+
+        if token is None and slack_conn_id is None:
+            self.log.error('No valid Slack token nor slack_conn_id supplied.')
+            raise AirflowException('No valid Slack token nor slack_conn_id supplied.')
+        if token is not None and slack_conn_id is not None:
+            self.log.error('Cannot determine Slack credential when both token and slack_conn_id are supplied.')
```
Comment (aoen) [inline airflow/operators/slack_operator.py]: Nit: No need to log the error (here and above), can just throw an exception.

File: airflow/hooks/slack_hook.py [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+        else:
+            raise AirflowException('Cannot get token: No valid Slack token nor slack_conn_id supplied.')
+
+    def call(self, method, api_params):
+        sc = SlackClient(self.token)
+        rc = sc.api_call(method, **api_params)
+
+        if not rc['ok']:
+            msg = "Slack API call failed (%s)".format(rc['error'])
+            self.log.error(msg)
```
Comment (aoen) [inline airflow/hooks/slack_hook.py]: No need to log the error, can just throw an exception.

**Comment by aoen** [comment]
LGTM besides nits

**Pull request closed without merging**